Replace anyhow with structured errors

vmx opened the pull request:
The issue at https://github.com/ipld/libipld/issues/115 contains a discussion about error handling. This PR is an experiment in using structured, distinct errors instead of anyhow.

I mark it as draft, as I hope to get feedback. This is really just and idea, if someone has other thoughts, please post them here.

This PR might be easier to review with its individual commits.

Stebalien commented on src/error.rs:
```diff
@@ -85,19 +24,15 @@ pub enum TypeErrorType {
 }
 
-impl From<Ipld> for TypeErrorType {
+impl From<Ipld> for KindErrorType {
```
I'd consider removing these. It kind of makes sense, but it could be dangerous given automatic error conversion with `?`.

I.e., turn this into `KindErrorType::from_ipld` function.

Stebalien commented on src/ipld.rs:
```diff
@@ -11,7 +11,20 @@ use core::fmt;
+
+/// Error when accessing IPLD List or Map elements.
+#[derive(Clone, Debug)]
+pub struct AccessError(String);
```
Can we use a non-exhaustive enum here instead?

Stebalien commented on src/error.rs:
```diff
@@ -1,72 +1,11 @@
+/// It maps the IPLD kinds into a unity-only enum.
 #[derive(Clone, Debug)]
-pub enum TypeErrorType {
+pub enum KindErrorType {
```
Do we need this to be error specific? Or can we just have a general `Kind` enum?

vmx replied:
That's a good idea!

Stebalien commented on src/serde/mod.rs:
```diff
@@ -8,6 +8,30 @@ mod ser;
 
+/// Error during Serde operations.
+#[derive(Clone, Debug)]
+pub struct SerdeError(String);
```
Ditto: it would be nice to have an enum.

Stebalien commented on src/convert.rs:
```diff
@@ -32,17 +81,21 @@ impl TryFrom<Ipld> for () {
+                    Ipld::$enum(value) => Ok(Some(value.try_into().map_err(|_| {
+                        ConversionError::FromIpld {
+                            from: KindErrorType::$enum,
+                            into: stringify!($yt),
```
ty?

Stebalien commented on src/convert.rs:
```diff
@@ -6,24 +6,73 @@ use alloc::{
+        /// The IPLD kind trying to convert from.
+        from: KindErrorType,
+        /// The type trying to convert into.
+        into: &'static str,
```
Can we not use kinds here as well?

Also, I'm not entirely sure how this error and the previous ones differ.

vmx replied:
This error is created by the `derive_try_from_*` macros. It's for converting from `Ipld` into native Rust types. Here's e.g. the call from `Ipld::Float` to and `f64`:  https://github.com/ipld/rust-ipld-core/blob/6d70de0114daa15e4062b57c466fbc5875ed3c64/src/convert.rs#L204

The `into` field is a string, so that the error can contain which native Rust type we tried to convert into. So it's not an Ipld Kind.

Stebalien replied:
Ah. Maybe we can use a TypeID? Although I'm not sure if that works (or is even worth it).

vmx replied:
Oh, I haven't thought of `TypeId`, that's even nicer (and more expressive than just  as string).

vmx commented:
Thanks @Stebalien for the review. I've updated the PR. Should the `Conversion` error type also be non-exhaustive?

It sounds like you're generally OK with having 3 different error types instead of a single catch-all enum, is that right?

Stebalien commented:
> Should the Conversion error type also be non-exhaustive?

IMO, we should default to non-exhaustive everywhere.

Stebalien commented:
> It sounds like you're generally OK with having 3 different error types instead of a single catch-all enum, is that right?

Absolutely!

Stebalien commented:
> IMO, we should default to non-exhaustive everywhere.

Of course, unless we're _sure_ that it's actually exhaustive...

Stebalien commented on src/ipld.rs:
```diff
@@ -11,7 +12,24 @@ use core::fmt;
+#[non_exhaustive]
+pub enum AccessError {
+    /// Error message describing the error.
+    Message(String),
```
I was actually hoping to enumerate the cases rather than have a generic "Message". Is that possible, or just too much work.

Personally, I'm not a huge fan of catch-all "message"/"custom"/"other" error variants (although I understand that they're sometimes required).

vmx replied:
I'll look if I find any structure. I'm more of a fan of such error messages as they can really be specific to what's going on. And in this case (accessing some field) it's really kind of a single error case, with a nice description.

Stebalien replied:
I agree we need something descriptive, but I prefer being able to programatically handle errors where possible. Of course, when we're dealing with malformed data... maybe this doesn't matter as much.

vmx replied:
Your message and my new push just crossed :)

vmx commented:
@Stebalien Please have another look. I agree, the structured `AccessError` makes things better.

I'll have a look at the Serde error tomorrow, whether there's a good way to make it also more structured.

Stebalien commented on src/ipld.rs:
```diff
@@ -16,14 +15,17 @@ use cid::Cid;
 #[derive(Clone, Debug)]
 #[non_exhaustive]
 pub enum AccessError {
```
nit: I'd call this `IndexError`, or something like that. For me, "access error" indicates that I'm trying to "access" some service.

vmx replied:
That's a good one, thanks!

Stebalien commented on src/ipld.rs:
```diff
@@ -66,7 +86,55 @@ impl fmt::Debug for Ipld {
+    ///
+    /// This is intentionally not implemented via `From<Ipld>` to prevent accidental conversions by
+    /// making it more explicit.
+    pub fn from_ipld(ipld: &Ipld) -> Self {
```
So... in retrospect, maybe `fn kind(&self)` on `Ipld` instead?

vmx replied:
Interesting, I had the same thought when I was coding this and was undecided, so I went for your proposal. Looks like I'll switch :)

Stebalien commented on src/ipld.rs:
```diff
@@ -94,53 +162,71 @@ impl<'a> From<&'a str> for IpldIndex<'a> {
+    fn try_from(index: IpldIndex<'a>) -> Result<Self, Self::Error> {
+        let parsed = match index {
+            IpldIndex::List(i) => i,
+            IpldIndex::Map(ref key) => key
+                .parse()
+                .map_err(|_| AccessError::ParseInteger(key.to_string()))?,
```
Why do we allow indexing arrays by integers anyways?

Stebalien replied:
And vice-versa? Honestly, that's more important because it relates to the whole "can we have integer keys" debate.

Stebalien replied:
NOTE: not relevant for _this_ PR. I can file an issue if you'd prefer to discuss there.

vmx replied:
Map keys are still always strings. Though I'll revisit this whole things in a subsequent PR.

Stebalien commented on src/serde/mod.rs:
```diff
@@ -5,9 +5,41 @@
+/// Error during Serde operations.
+#[derive(Clone, Debug)]
+#[non_exhaustive]
+pub enum SerdeError {
+    /// Error message describing the error.
+    Message(String),
+}
```
If this can only be a string message (it looks like that might be the case), I take back what I said earlier. There's no need to make this an enum in that case (we won't have any other cases, unless I'm mistaken).

My previous point was about preserving information by representing errors as structured data when we have enough information to do so. If we only have a string, we're SOL anyways.

vmx replied:
I've split it into serialization and deserialization errors. More structure didn't make sense to me, as the errors are quite individual. The only ones could be the ones not matching the right Ipld Kind, but even there, I like such detailed error messages more and I don't feel like there would be any need to programatically catch them in a more structured way.

Stebalien replied:
Honestly, I don't think deserialize versus serialize is important to distinguish in this case (the user can just annotate the error if they care). And/or, if we _did_ want to distinguish those cases, I'd just have two errors.

But again, I'm just quibbling at this point.

vmx replied:
Good point, usually you know whether you are serializing or deserializing.

Stebalien reviewed (approved):
Ok, I'm just quibbling at this point so I'm going to approve this.

vmx commented:
> Ok, I'm just quibbling at this point so I'm going to approve this.

No. Your feedback is really welcome. I want to get this done properly. So please keep going.

I've addressed your points, I think it's good to go now (I know you already approved it, but another re-approval would be appreciated).

vmx commented:
I've moved back to having the SerdeError as newtype struct wrapping a `String`. The only additional thing is that I fixed tests that were broken due to earlier changes.

I merge this PR as it was already approved. In case there's disagreement on changes later on, things can still be changed before a new release it cut.
